rclcpp/include/rclcpp/publisher_options.hpp

```diff
@@ -75,7 +75,9 @@ struct PublisherOptionsWithAllocator : public PublisherOptionsBase
     using AllocatorTraits = std::allocator_traits<Allocator>;
     using MessageAllocatorT = typename AllocatorTraits::template rebind_alloc<MessageT>;
     auto message_alloc = std::make_shared<MessageAllocatorT>(*this->get_allocator().get());
-    result.allocator = rclcpp::allocator::get_rcl_allocator<MessageT>(*message_alloc);
+    // TODO(stevewolter): This is likely to invoke undefined behavior - message_alloc goes
+    // out of scope at the end of this function, but the allocator doesn't.
+    result.allocator = get_rcl_allocator(*message_alloc);
```

Comment:
Can you open an issue about this?  It's a good thing to have it in the code, but it is easy to forget about.

Reply:
Done, #1339.

Reply:
Actually, I does this hit UB? Since we're doing a copy of `MessageAllocatorT`, it should be fine. According to the expected behavior of allocators, they should be trivially copyable:

- https://en.cppreference.com/w/cpp/named_req/Allocator#Requirements
- https://github.com/ros2/rclcpp/issues/849#issuecomment-539758462

So since we're copying it, the allocator instance in `message_alloc` should be fine to go out of scope.

There's no reason for us to be using shared pointers to allocators, which is something that's already been pointed out but we've yet to change. Instead we should just be copying the allocators around.

Reply:
Note that we're *not* copying the allocator in the generic overload of get_rcl_allocator, we're taking its address and pass it by reference. There's no way to copy it in the context of get_rcl_allocator, because the state pointer in rcl_allocator_t is a non-owning pointer (because there's no way to specify a destructor in rcl_allocator_t).

I think we can get away with it for now (both allocators in ROS are stateless), but the move to polymorphic allocators will need to resolve the issue.

Reply:
Ah you're right, the `void * state` field in the allocator structure assumes essentially global state. For a generic allocator there's no way to deal with this, but if your allocator exposes its global state you could store a pointer to it in the state field and reconstruct the allocator each time you wanted to do that.

In fact that's really what we should be doing rather than storing a reference to the allocator, we should just be capturing the allocator type and creating a new instance each time we need to do an operation, using a lambda or something.

But that all being said, we need to just move away from these template based allocators I think.

----

For this particular issue we could keep the `message_alloc` as a member of the options class and at least that would extend the validity of the `rcl_publisher_options_t result` we're returning to match the lifetime of the publisher options object (which is itself stored in the publisher):

https://github.com/ros2/rclcpp/blob/869f3ed873c0cdd0bd179ce7776676492a386e41/rclcpp/include/rclcpp/publisher.hpp#L373

So that should be safe for the typical use case.